source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
-                            ptr -= alloc_sector;
-                            //Set return Pointer
-                            retval = (void *) ptr;
-                            ptr--;
-                            //Set End length
-                            *ptr-- = alloc_sector;
-                            //Now set new slice start & End len fields
-                            //Move pointer to New Free Sector
-                            //Calculate new sector size
-                            alloc_sector += 2; //
-
-                            s_size = (s_size - alloc_sector);
-
-                            *ptr-- = -(s_size);
-                            ptr -= s_size;
-                            *ptr = -(s_size);
-                            if (mem_stat_info_ptr) {
-                                alloc_sector -= 2;
-                            }
+                        if (block_total_size > (alloc_block + 4)) {
+                            //There is enough room for a new hole so create it first
+                            int *hole_ptr = ptr + ((alloc_block + 2) * direction);
+                            int hole_size = block_total_size - alloc_block - 2;
+                            *hole_ptr = -(hole_size);
+                            hole_ptr += (hole_size + 1) * direction;
+                            *hole_ptr = -(hole_size);
+                        }else{
+                            alloc_block = block_total_size;
+                        }
+
+                        *ptr = alloc_block;
+                        if (direction > 0) {
+                            ptr++;
+                            retval = ptr;
+                            ptr += alloc_block;
                         } else {
-                            //debug("USE Same\r\n");
-                            //Set sector start len & Move pointer to Allocated data sector
-                            *ptr = s_size;
-                            ptr -= s_size;
-                            retval = (void *) ptr;
+                            ptr -= alloc_block;
+                            retval = ptr;
                             ptr--;
-                            //Set End length
-                            *ptr-- = s_size;
                         }
+                        *ptr = alloc_block;
                         break;
-                    } else {
-                        //Set Number from negative to Positive
-                        s_size = -s_size;
                     }
                 }
-                //Move sector to next
-                //Allocated sector
-                moved += (s_size + 2);
-                ptr -= 2; //Skip Both of length fileds
-                ptr -= s_size;
-                if (heap_main == ptr) {
-                    break;
-                }
+
+                moved += (block_total_size + 2);
+                ptr += direction * (block_total_size + 2);
```

Comment:
Still being inconsistent about whether multiply by direction is good or bad. I don't have a firm view, but stick to one style (compare ns_block_validate).